



<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>DeepExplainer Genomics Example &mdash; SHAP latest documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=7ab3649f" />
      <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />
      <link rel="stylesheet" href="../../_static/css/style.css" type="text/css" />

  
    <link rel="shortcut icon" href="../../_static/favicon.ico"/>
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=c6e86fd7"></script>
      <script src="../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="API Reference" href="../../api.html" />
    <link rel="prev" title="Genomic examples" href="../../genomic_examples.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #343131" >

          
          
          <a href="../../index.html">
            
              <img src="../../_static/shap_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../overviews.html">Topical overviews</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../tabular_examples.html">Tabular examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../text_examples.html">Text examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../image_examples.html">Image examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../genomic_examples.html">Genomic examples</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">DeepExplainer Genomics Example</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Obtain-data-and-keras-model">Obtain data and keras model</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Compute-importance-scores">Compute importance scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Visualize-scores-on-individual-sequences">Visualize scores on individual sequences</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api_examples.html">API examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../benchmarks.html">Benchmarks</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #343131" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">SHAP</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../genomic_examples.html">Genomic examples</a></li>
      <li class="breadcrumb-item active">DeepExplainer Genomics Example</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/example_notebooks/genomic_examples/DeepExplainer Genomics Example.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="DeepExplainer-Genomics-Example">
<h1>DeepExplainer Genomics Example<a class="headerlink" href="#DeepExplainer-Genomics-Example" title="Link to this heading"></a></h1>
<p>This runs DeepExplainer with the model trained on simulated genomic data from the DeepLIFT repo (<a class="reference external" href="https://github.com/kundajelab/deeplift/blob/master/examples/genomics/genomics_simulation.ipynb">https://github.com/kundajelab/deeplift/blob/master/examples/genomics/genomics_simulation.ipynb</a>), using a dynamic reference (i.e. the reference varies depending on the input sequence; in this case, the reference is a collection of dinucleotide-shuffled versions of the input sequence)</p>
<p>The simulated data was as follows:</p>
<ul class="simple">
<li><p>1/4 sequences with BOTH 1-3 instances of a GATA_disc1 motif AND 1-3 instances of a TAL1_known1 motif; these were labelled 1,1,1</p></li>
<li><p>1/4 sequences with 1-3 instances of a GATA_disc1 motif embedded; these were labelled 0,1,0</p></li>
<li><p>1/4 sequences with 1-3 instances of a TAL1_known1 motif embedded; these were labelled 0,0,1</p></li>
<li><p>1/4 sequences with no motif; these were labelled 0,0,0</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
<section id="Obtain-data-and-keras-model">
<h2>Obtain data and keras model<a class="headerlink" href="#Obtain-data-and-keras-model" title="Link to this heading"></a></h2>
<p>Pull in the relevant data</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span><span class="o">[[</span><span class="w"> </span>!<span class="w"> </span>-f<span class="w"> </span>sequences.simdata.gz<span class="w"> </span><span class="o">]]</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span>wget<span class="w"> </span>https://raw.githubusercontent.com/AvantiShri/model_storage/db919b12f750e5844402153233249bb3d24e9e9a/deeplift/genomics/sequences.simdata.gz
<span class="o">!</span><span class="w"> </span><span class="o">[[</span><span class="w"> </span>!<span class="w"> </span>-f<span class="w"> </span>keras2_conv1d_record_5_model_PQzyq_modelJson.json<span class="w"> </span><span class="o">]]</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span>wget<span class="w"> </span>https://raw.githubusercontent.com/AvantiShri/model_storage/b6e1d69/deeplift/genomics/keras2_conv1d_record_5_model_PQzyq_modelJson.json
<span class="o">!</span><span class="w"> </span><span class="o">[[</span><span class="w"> </span>!<span class="w"> </span>-f<span class="w"> </span>keras2_conv1d_record_5_model_PQzyq_modelWeights.h5<span class="w"> </span><span class="o">]]</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span>wget<span class="w"> </span>https://raw.githubusercontent.com/AvantiShri/model_storage/b6e1d69/deeplift/genomics/keras2_conv1d_record_5_model_PQzyq_modelWeights.h5
<span class="o">!</span><span class="w"> </span><span class="o">[[</span><span class="w"> </span>!<span class="w"> </span>-f<span class="w"> </span>test.txt.gz<span class="w"> </span><span class="o">]]</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span>wget<span class="w"> </span>https://raw.githubusercontent.com/AvantiShri/model_storage/9aadb769735c60eb90f7d3d896632ac749a1bdd2/deeplift/genomics/test.txt.gz
</pre></div>
</div>
</div>
<p>Load the data</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>simdna
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Requirement already satisfied: simdna in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (0.4.3.2)
Requirement already satisfied: numpy&gt;=1.9 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from simdna) (1.26.3)
Requirement already satisfied: matplotlib in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from simdna) (3.8.2)
Requirement already satisfied: scipy in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from simdna) (1.11.4)
Requirement already satisfied: contourpy&gt;=1.0.1 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (1.2.0)
Requirement already satisfied: cycler&gt;=0.10 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (0.12.1)
Requirement already satisfied: fonttools&gt;=4.22.0 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (4.47.2)
Requirement already satisfied: kiwisolver&gt;=1.3.1 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (1.4.5)
Requirement already satisfied: packaging&gt;=20.0 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (23.2)
Requirement already satisfied: pillow&gt;=8 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (10.2.0)
Requirement already satisfied: pyparsing&gt;=2.3.1 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (3.1.1)
Requirement already satisfied: python-dateutil&gt;=2.7 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from matplotlib-&gt;simdna) (2.8.2)
Requirement already satisfied: six&gt;=1.5 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from python-dateutil&gt;=2.7-&gt;matplotlib-&gt;simdna) (1.16.0)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">gzip</span>

<span class="kn">import</span> <span class="nn">simdna.synthetic</span> <span class="k">as</span> <span class="nn">synthetic</span>

<span class="n">data_filename</span> <span class="o">=</span> <span class="s2">&quot;sequences.simdata.gz&quot;</span>

<span class="c1"># read in the data in the testing set</span>
<span class="n">test_ids_fh</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;test.txt.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="n">ids_to_load</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">test_ids_fh</span><span class="p">]</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">synthetic</span><span class="o">.</span><span class="n">read_simdata_file</span><span class="p">(</span><span class="n">data_filename</span><span class="p">,</span> <span class="n">ids_to_load</span><span class="o">=</span><span class="n">ids_to_load</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="c1"># this is set up for 1d convolutions where examples</span>
<span class="c1"># have dimensions (len, num_channels)</span>
<span class="c1"># the channel axis is the axis for one-hot encoding.</span>
<span class="k">def</span> <span class="nf">one_hot_encode_along_channel_axis</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
    <span class="n">to_return</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
    <span class="n">seq_to_one_hot_fill_in_array</span><span class="p">(</span><span class="n">zeros_array</span><span class="o">=</span><span class="n">to_return</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span> <span class="n">one_hot_axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">to_return</span>


<span class="k">def</span> <span class="nf">seq_to_one_hot_fill_in_array</span><span class="p">(</span><span class="n">zeros_array</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">one_hot_axis</span><span class="p">):</span>
    <span class="k">assert</span> <span class="n">one_hot_axis</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">one_hot_axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">zeros_array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">one_hot_axis</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">zeros_array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="c1"># will mutate zeros_array</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">char</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">char</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">):</span>
            <span class="n">char_idx</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">elif</span> <span class="n">char</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">):</span>
            <span class="n">char_idx</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">char</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">):</span>
            <span class="n">char_idx</span> <span class="o">=</span> <span class="mi">2</span>
        <span class="k">elif</span> <span class="n">char</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">):</span>
            <span class="n">char_idx</span> <span class="o">=</span> <span class="mi">3</span>
        <span class="k">elif</span> <span class="n">char</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">,</span> <span class="s2">&quot;n&quot;</span><span class="p">):</span>
            <span class="k">continue</span>  <span class="c1"># leave that pos as all 0&#39;s</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Unsupported character: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">char</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">one_hot_axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">zeros_array</span><span class="p">[</span><span class="n">char_idx</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">one_hot_axis</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">zeros_array</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">char_idx</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>


<span class="n">onehot_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">one_hot_encode_along_channel_axis</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>Load the model</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">keras.models</span> <span class="kn">import</span> <span class="n">model_from_json</span>

<span class="c1"># load the keras model</span>
<span class="n">keras_model_weights</span> <span class="o">=</span> <span class="s2">&quot;keras2_conv1d_record_5_model_PQzyq_modelWeights.h5&quot;</span>
<span class="n">keras_model_json</span> <span class="o">=</span> <span class="s2">&quot;keras2_conv1d_record_5_model_PQzyq_modelJson.json&quot;</span>

<span class="n">keras_model</span> <span class="o">=</span> <span class="n">model_from_json</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">keras_model_json</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="n">keras_model</span><span class="o">.</span><span class="n">load_weights</span><span class="p">(</span><span class="n">keras_model_weights</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Install the deeplift package for the dinucleotide shuffling and visualization code</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>deeplift
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Requirement already satisfied: deeplift in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (0.6.13.0)
Requirement already satisfied: numpy&gt;=1.9 in /home/xiao/miniforge3/envs/shap/lib/python3.11/site-packages (from deeplift) (1.26.3)
</pre></div></div>
</div>
</section>
<section id="Compute-importance-scores">
<h2>Compute importance scores<a class="headerlink" href="#Compute-importance-scores" title="Link to this heading"></a></h2>
<p>Define the function that generates the reference, in this case by performing a dinucleotide shuffle of the given input sequence</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">deeplift.dinuc_shuffle</span> <span class="kn">import</span> <span class="n">dinuc_shuffle</span>


<span class="k">def</span> <span class="nf">shuffle_several_times</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
    <span class="n">s</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">dinuc_shuffle</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">num_shufs</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Run DeepExplainer with the dynamic reference function</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">shap</span>

<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

<span class="n">seqs_to_explain</span> <span class="o">=</span> <span class="n">onehot_data</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">]]</span>  <span class="c1"># these three are positive for task 0</span>
<span class="n">dinuc_shuff_explainer</span> <span class="o">=</span> <span class="n">shap</span><span class="o">.</span><span class="n">DeepExplainer</span><span class="p">((</span><span class="n">keras_model</span><span class="o">.</span><span class="n">input</span><span class="p">,</span> <span class="n">keras_model</span><span class="o">.</span><span class="n">output</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]),</span> <span class="n">shuffle_several_times</span><span class="p">)</span>
<span class="n">raw_shap_explanations</span> <span class="o">=</span> <span class="n">dinuc_shuff_explainer</span><span class="o">.</span><span class="n">shap_values</span><span class="p">(</span><span class="n">seqs_to_explain</span><span class="p">,</span> <span class="n">check_additivity</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Visualize-scores-on-individual-sequences">
<h2>Visualize scores on individual sequences<a class="headerlink" href="#Visualize-scores-on-individual-sequences" title="Link to this heading"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">deeplift.visualization</span> <span class="kn">import</span> <span class="n">viz_sequence</span>

<span class="c1"># project the importance at each position onto the base that&#39;s actually present</span>
<span class="n">dinuc_shuff_explanations</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">raw_shap_explanations</span><span class="p">,</span> <span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)[:,</span> <span class="p">:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">seqs_to_explain</span>
<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">dinuc_shuff_explanation</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="n">dinuc_shuff_explanations</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Scores for example&quot;</span><span class="p">,</span> <span class="n">idx</span><span class="p">)</span>
    <span class="n">highlight</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;blue&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">(</span><span class="n">embedding</span><span class="o">.</span><span class="n">startPos</span><span class="p">,</span> <span class="n">embedding</span><span class="o">.</span><span class="n">startPos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">embedding</span><span class="o">.</span><span class="n">what</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">embedding</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">embeddings</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;GATA_disc1&quot;</span> <span class="ow">in</span> <span class="n">embedding</span><span class="o">.</span><span class="n">what</span><span class="o">.</span><span class="n">getDescription</span><span class="p">()</span>
        <span class="p">],</span>
        <span class="s2">&quot;green&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">(</span><span class="n">embedding</span><span class="o">.</span><span class="n">startPos</span><span class="p">,</span> <span class="n">embedding</span><span class="o">.</span><span class="n">startPos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">embedding</span><span class="o">.</span><span class="n">what</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">embedding</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">embeddings</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;TAL1_known1&quot;</span> <span class="ow">in</span> <span class="n">embedding</span><span class="o">.</span><span class="n">what</span><span class="o">.</span><span class="n">getDescription</span><span class="p">()</span>
        <span class="p">],</span>
    <span class="p">}</span>
    <span class="n">viz_sequence</span><span class="o">.</span><span class="n">plot_weights</span><span class="p">(</span><span class="n">dinuc_shuff_explanation</span><span class="p">,</span> <span class="n">subticks_frequency</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">highlight</span><span class="o">=</span><span class="n">highlight</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scores for example 0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/example_notebooks_genomic_examples_DeepExplainer_Genomics_Example_21_1.png" src="../../_images/example_notebooks_genomic_examples_DeepExplainer_Genomics_Example_21_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scores for example 3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/example_notebooks_genomic_examples_DeepExplainer_Genomics_Example_21_3.png" src="../../_images/example_notebooks_genomic_examples_DeepExplainer_Genomics_Example_21_3.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scores for example 9
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/example_notebooks_genomic_examples_DeepExplainer_Genomics_Example_21_5.png" src="../../_images/example_notebooks_genomic_examples_DeepExplainer_Genomics_Example_21_5.png" />
</div>
</div>
<p>The plots above show the importance scores for three example sequences on the task that predicts sequences containing both GATA_disc1 and TAL1_known1 motifs. Letter height reflects the score. Blue boxes indicate the ground-truth locations of the inserted GATA_disc1 motifs, green boxes indicate the ground-truth locations of the inserted TAL1_known1 motifs.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../../genomic_examples.html" class="btn btn-neutral float-left" title="Genomic examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../../api.html" class="btn btn-neutral float-right" title="API Reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2018, Scott Lundberg.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>